### PR TITLE
cacher.go: remove NewCacher func

### PR DESCRIPTION
**Pull request by xiang90**
NewCacher is a wrapper of NewCacherFromConfig. NewCacher understands
how to create a key func from scopeStrategy. However, it is not the
responsibility of cacher. So we should remove this function, and
construct the config in its caller, which should understand scopeStrategy.

### CODE REVIEW

File: pkg/registry/generic/registry/storage_factory.go [hunk 1]
```diff
@@ -31,7 +31,24 @@ func StorageWithCacher(
...
-	return storage.NewCacher(
-		storageInterface, capacity, etcdstorage.APIObjectVersioner{},
-		objectType, resourcePrefix, scopeStrategy, newListFunc)
+
+	config := storage.CacherConfig{
+		CacheCapacity:  capacity,
+		Storage:        storageInterface,
+		Versioner:      etcdstorage.APIObjectVersioner{},
+		Type:           objectType,
+		ResourcePrefix: resourcePrefix,
```
Comment (liggitt) [inline pkg/registry/generic/registry/storage_factory.go]: Would it be possible to remove resourcePrefix from the cacher as well? That doesn't seem like something the caching layer should need to know

Comment (xiang90) [inline pkg/registry/generic/registry/storage_factory.go]: it is possible I feel. 

But can we try to do that in another pull request? It requires more thought. My general feeling was the cacher is a little bit complicated right now. I am trying to simplify it or at least make it easier to read.

**Comment by liggitt** [comment]
this refactor looks fine.

I'd like to see additional simplifications as follow ups - 

1. cacher assumes a standard etcd key construction, which makes customization of KeyFunc in the various registries impossible. The existing ObjectNameFunc/KeyFunc should probably be used by the cacher to determine the etcd key of an object

2. cacher assumes a single etcd prefix, which makes customization of KeyRootFunc in the various registries impossible. The existing KeyRootFunc should probably be used by the cacher to determine the applicable prefix for a request

**Comment by xiang90** [comment]
@liggitt Thanks. I will resolve the issues you mentioned in follow-up prs.
